engine/src/main/java/org/terasology/entitySystem/entity/internal/PojoEntityCache.java

```diff
@@ -0,0 +1,350 @@
+/*
+ * Copyright 2017 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.entitySystem.entity.internal;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.MapMaker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.entitySystem.Component;
+import org.terasology.entitySystem.entity.EntityBuilder;
+import org.terasology.entitySystem.entity.EntityCache;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.entity.lifecycleEvents.BeforeDeactivateComponent;
+import org.terasology.entitySystem.entity.lifecycleEvents.BeforeEntityCreated;
+import org.terasology.entitySystem.entity.lifecycleEvents.BeforeRemoveComponent;
+import org.terasology.entitySystem.entity.lifecycleEvents.OnActivatedComponent;
+import org.terasology.entitySystem.entity.lifecycleEvents.OnAddedComponent;
+import org.terasology.entitySystem.event.internal.EventSystem;
+import org.terasology.entitySystem.prefab.Prefab;
+import org.terasology.logic.location.LocationComponent;
+import org.terasology.math.geom.Quat4f;
+import org.terasology.math.geom.Vector3f;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static org.terasology.entitySystem.entity.internal.PojoEntityManager.NULL_ID;
+
+/**
+ */
+public class PojoEntityCache implements EntityCache {
+
+    private PojoEntityManager entityManager;
+
+    private static final Logger logger = LoggerFactory.getLogger(PojoEntityCache.class);
+
+    private Map<Long, BaseEntityRef> entityStore = new MapMaker().weakValues().concurrencyLevel(4).initialCapacity(1000).makeMap();
```

Comment:
I hadn't considered entities being in multiple caches. It might be useful (e.g. for entities near a sector boundary), but it could cause lots of problems with systems trying to update the same entity at the same time in two different caches (especially if we're thinking about using multi-threading to handle sectors in the future). So I think it's best to have each entity in a single cache.

I think that'll definitely be necessary when we have multiple sectors, so sectors can be moved or split (and entities can change sector if they move around, or find themselves interacting with lots of things outside of their sector).